common/djangoapps/student/models.py

```diff
@@ -1894,29 +1894,64 @@ def refundable(self, user_already_has_certs_for=None):
     def refund_cutoff_date(self):
         """ Calculate and return the refund window end date. """
         # NOTE: This is here to avoid circular references
-        from openedx.core.djangoapps.commerce.utils import ecommerce_api_client, ECOMMERCE_DATE_FORMAT
-
+        from openedx.core.djangoapps.commerce.utils import ECOMMERCE_DATE_FORMAT
         date_placed = self.get_order_attribute_value('date_placed')
 
         if not date_placed:
             order_number = self.get_order_attribute_value('order_number')
             if not order_number:
                 return None
 
+            date_placed = self.get_order_attribute_from_ecommerce('date_placed')
+            if not date_placed:
+                return None
+
+            # also save the attribute so that we don't need to call ecommerce again.
+            username = self.user.username
+            enrollment_attributes = get_enrollment_attributes(username, str(self.course_id))
+            enrollment_attributes.append(
+                {
+                    "namespace": "order",
+                    "name": "date_placed",
+                    "value": date_placed,
+                }
+            )
+            set_enrollment_attributes(username, str(self.course_id), enrollment_attributes)
+
+        refund_window_start_date = max(
+            datetime.strptime(date_placed, ECOMMERCE_DATE_FORMAT),
+            self.course_overview.start.replace(tzinfo=None)
+        )
+
+        return refund_window_start_date.replace(tzinfo=UTC) + EnrollmentRefundConfiguration.current().refund_window
+
+    def is_order_voucher_refundable(self):
+        """ Checks if the coupon batch expiration date has passed to determine whether order voucher is refundable. """
+        from openedx.core.djangoapps.commerce.utils import ECOMMERCE_DATE_FORMAT
+        vouchers = self.get_order_attribute_from_ecommerce('vouchers')
+        if not vouchers:
+            return False
+        voucher_end_datetime_str = vouchers[0]['end_datetime']
+        voucher_expiration_date = datetime.strptime(voucher_end_datetime_str, ECOMMERCE_DATE_FORMAT).replace(tzinfo=UTC)
+        return datetime.now(UTC) < voucher_expiration_date
```

Comment:
Not a blocker but it would be great to add this logic in ecommerce at [here](https://github.com/edx/ecommerce/blob/76044029feae4d3709cd4a92c0118adc6934851f/ecommerce/extensions/api/serializers.py#L757)

Reply:
Agreed.